src/main/java/edu/illinois/library/cantaloupe/image/exif/Reader.java

```diff
@@ -157,6 +157,10 @@ private byte[] readBytes(int length) throws IOException {
         byte[] data = new byte[length];
         int n, offset = 0;
         while ((n = inputStream.read(data, offset, data.length - offset)) < offset) {
+            if (n == -1) {
+                LOGGER.trace("readBytes(): No more data to read");
+                break;
+            }
             offset += n;
         }
```

Comment:
Have recently stumbled across a similar implementation in [4.1's `JPEG2000MetadataReader`](https://github.com/cantaloupe-project/cantaloupe/blob/release/4.1/src/main/java/edu/illinois/library/cantaloupe/processor/codec/JPEG2000MetadataReader.java#L409-L417) which was breaking due to _not_ reading the full structure. We're presently moving to patch it, with something more like:

```suggestion
        while ((n = inputStream.read(data, offset, data.length - offset)) > -1) {
            offset += n;
            if (offset >= length) {
              break;
            }
        }
        if (offset < length) {
          throw new IOException("Failed to read the expected amount of data.");
        }
```

Reply:
@glenrobson: I've thrown together https://github.com/cantaloupe-project/cantaloupe/pull/659 making use of `inputStream.readFully()` here, and a few other places, which I think should do the trick for this and a couple of other places.